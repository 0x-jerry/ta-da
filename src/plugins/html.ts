import { type TypeItem, TadaItemType, type TadaRenderer } from '../core'

export function createHtmlRenderer() {
  const opt = {
    el: null as null | HTMLElement,
    cache: '',
  }

  const core: TadaRenderer = {
    split(str) {
      const items: TypeItem[] = []

      const d = document.createElement('div')
      d.innerHTML = str

      d.childNodes.forEach((item) => splitHtml(item, items))

      return items
    },
    render(item) {
      if (!opt.el) return

      opt.cache += item.content

      opt.el.innerHTML = opt.cache
    },
    clear() {
      opt.cache = ''
    },
  }

  const renderer = {
    setup(el: HTMLElement) {
      opt.el = el
    },
  }

  return {
    ...core,
    ...renderer,
  }
}

function splitHtml(node: Node, items: TypeItem[]) {
  if (node.nodeType === document.TEXT_NODE) {
    const nodes = (node.textContent || '').split('').map((n) => {
      return {
        type: /\s/.test(n) ? TadaItemType.Space : TadaItemType.Text,
        content: n,
      }
    })

    items.push(...nodes)
    return
  }

  if (node.nodeType === document.ELEMENT_NODE) {
    items.push({
      type: TadaItemType.Invisible,
      content: `<${node.nodeName}>`,
    })

    node.childNodes.forEach((item) => splitHtml(item, items))

    items.push({
      type: TadaItemType.Invisible,
      content: `</${node.nodeName}>`,
    })

    return
  }
}
