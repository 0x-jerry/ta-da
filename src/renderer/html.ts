import { ensureArray } from '@0x-jerry/utils'
import { type TadaItem, type TadaRenderer, TadaItemType } from '../types'
import { splitText } from './_utils'

export function createHtmlRenderer() {
  const opt = {
    el: null as null | HTMLElement,
    cache: '',
  }

  const core: TadaRenderer = {
    split(str) {
      const items: TadaItem[] = []

      const parsedDom = document.createElement('div')
      parsedDom.innerHTML = str

      parsedDom.childNodes.forEach((item) => splitHtml(item, items))

      return items
    },
    render(items) {
      for (const item of ensureArray(items)) {
        opt.cache += item.content
      }

      if (!opt.el) return
      opt.el.innerHTML = opt.cache
    },
    reset() {
      opt.cache = ''
    },
  }

  const renderer = {
    setup(el: HTMLElement) {
      opt.el = el
    },
  }

  return {
    ...core,
    ...renderer,
  }
}

function splitHtml(node: Node, items: TadaItem[]) {
  if (node.nodeType === document.TEXT_NODE) {
    const nodes = splitText(node.textContent || '').map((n) => {
      return {
        type: /\s/.test(n) ? TadaItemType.Space : TadaItemType.Text,
        content: n,
      }
    })

    items.push(...nodes)
    return
  }

  if (node.nodeType === document.ELEMENT_NODE) {
    items.push({
      type: TadaItemType.Invisible,
      content: `<${node.nodeName}>`,
    })

    node.childNodes.forEach((item) => splitHtml(item, items))

    items.push({
      type: TadaItemType.Invisible,
      content: `</${node.nodeName}>`,
    })

    return
  }
}
