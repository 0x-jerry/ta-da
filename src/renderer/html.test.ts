// @vitest-environment happy-dom

import { createHtmlRenderer } from './html'

describe('terminal renderer', () => {
  it('should split html correctly', () => {
    const r = createHtmlRenderer()
    const items = r.split('<div> hello <b> world <em> eee </em>  world </b> ğŸš€ğŸš€ 123 </div>')

    expect(items).matchSnapshot()
  })
})
