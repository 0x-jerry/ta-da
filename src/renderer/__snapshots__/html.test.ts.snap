// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`terminal renderer > should split html correctly 1`] = `
[
  {
    "content": "<DIV>",
    "type": "invisible",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "h",
    "type": "text",
  },
  {
    "content": "e",
    "type": "text",
  },
  {
    "content": "l",
    "type": "text",
  },
  {
    "content": "l",
    "type": "text",
  },
  {
    "content": "o",
    "type": "text",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "<B>",
    "type": "invisible",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "w",
    "type": "text",
  },
  {
    "content": "o",
    "type": "text",
  },
  {
    "content": "r",
    "type": "text",
  },
  {
    "content": "l",
    "type": "text",
  },
  {
    "content": "d",
    "type": "text",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "<EM>",
    "type": "invisible",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "e",
    "type": "text",
  },
  {
    "content": "e",
    "type": "text",
  },
  {
    "content": "e",
    "type": "text",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "</EM>",
    "type": "invisible",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "w",
    "type": "text",
  },
  {
    "content": "o",
    "type": "text",
  },
  {
    "content": "r",
    "type": "text",
  },
  {
    "content": "l",
    "type": "text",
  },
  {
    "content": "d",
    "type": "text",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "</B>",
    "type": "invisible",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "ðŸš€",
    "type": "text",
  },
  {
    "content": "ðŸš€",
    "type": "text",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "1",
    "type": "text",
  },
  {
    "content": "2",
    "type": "text",
  },
  {
    "content": "3",
    "type": "text",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "</DIV>",
    "type": "invisible",
  },
]
`;
